### PR TITLE
[RDY] Add noinsert and noselect features in completeopt

### PR DESCRIPTION
**Pull request by Shougo**
I backported the patch from https://groups.google.com/forum/#!searchin/vim_dev/completeopt/vim_dev/tVsk0pdOGvs/fCzBbPkA4w0J.
It is already in todo list.  But it is not merged in long time.

Note: The original authors are me and @DeaR.
I got the permission from @DeaR.

Usecase: https://github.com/Shougo/neocomplcache.vim/issues/426

### CODE REVIEW

**Comment by syswow** [comment]
I want this feature. :smile_cat:

**Comment by justinmk** [comment]
Tests would be very helpful.

**Comment by Shougo** [comment]
@justinmk Unfortunately, I cannot find tests about completions in Vim/neovim...

**Comment by fmoralesc** [comment]
@Shougo I think you can model the test on some of the ui ones, for example  [tests/functional/ui/wildmode_spec.lua](https://github.com/neovim/neovim/blob/master/test/functional/ui/wildmode_spec.lua), where you make the test expect some specific contents in the screen after you feed it some inputs. That might help with noinsert, but I'm not sure about noselect...

**Comment by Shougo** [comment]
Rebased.

**Comment by Shougo** [comment]
@fmoralesc I added the completion spec file in https://github.com/neovim/neovim/pull/2563.
I will add the spec for it after #2563 is merged.

**Comment by HerringtonDarkholme** [comment]
Looking forward this being merged.

**Comment by Shougo** [comment]
#2563 must be merged in first.

**Comment by Shougo** [comment]
https://github.com/neovim/neovim/pull/2563

I will add the tests after #2563 is merged.
Please review it before adding the tests.

**Comment by fmoralesc** [comment]
@Shougo I did last night after reading your messages at gitter, and it looks good. The documentation text is misaligned, but other than that it looks good.

File: src/nvim/edit.c [hunk 1]
```diff
@@ -3972,7 +3989,10 @@ ins_compl_next (
   }
 
   /* Insert the text of the new completion, or the compl_leader. */
-  if (insert_match) {
+  if (compl_no_insert && !started) {
+    ins_bytes(compl_orig_text + ins_compl_len());
+    compl_used_match = FALSE;
+  } else if (insert_match) {
```
Comment (splinterofchaos) [inline src/nvim/edit.c]: Would it make sense to put something like
```c
if (compl_no_insert) {
  insert_match = false;
}
```
above this?

**Comment by splinterofchaos** [comment]
The `TRUE`/`FALSE`s should be lowercase, and this adds some c-style comments (`/* */`), but this looks good at a glance. I'll look more thoroughly at it after #2563.

**Comment by fwalch** [comment]
> The TRUE/FALSEs should be lowercase

The datatype of these variables should then also be changed from `int` to `bool`.

**Comment by fmoralesc** [comment]
@splinterofchaos, @fwalch Maybe those `bool`s could be fixed after merging this PR, to have a separate commit that fixes if for the other `compl_` flags (https://github.com/Shougo/neovim/blob/completeopt/src/nvim/edit.c#L148-L177)? I don't think we need to burden @Shougo with this.

**Comment by fwalch** [comment]
@fmoralesc :+1:

**Comment by splinterofchaos** [comment]
> I don't think we need to burden @Shougo with this.

Agreed, and I'm no stickler for that kinda thing anyway.

File: src/nvim/edit.c (hunk 1 above)
Comment (Shougo) [inline src/nvim/edit.c]: Oh, it seems good.  Thanks.

**Comment by Shougo** [comment]
> @splinterofchaos, @fwalch Maybe those bools could be fixed after merging this PR, to have a separate commit that fixes if for the other compl_ flags (https://github.com/Shougo/neovim/blob/completeopt/src/nvim/edit.c#L148-L177)? I don't think we need to burden @Shougo with this. 

Yes.  I think so too.  Because, the problem has in other flags(`compl_get_longest`, `compl_used_match`).
It should be fixed.

> The TRUE/FALSEs should be lowercase, and this adds some c-style comments (/* */), but this looks good at a glance. I'll look more thoroughly at it after #2563.

And the comments style also be fixed in other PR.

File: src/nvim/edit.c [hunk 2]
```diff
@@ -3997,7 +4017,10 @@ ins_compl_next (
 
   /* Enter will select a match when the match wasn't inserted and the popup
    * menu is visible. */
-  compl_enter_selects = !insert_match && compl_match_array != NULL;
+  if (compl_no_insert && !started)
+    compl_enter_selects = TRUE;
+  else
+    compl_enter_selects = !insert_match && compl_match_array != NULL;
```
Comment (Pyrohh) [inline src/nvim/edit.c]: This should have braces.

**Comment by Shougo** [comment]
Thanks.  I will update it for review and add the tests in the weekend.

File: src/nvim/edit.c (hunk 1 above)
Comment (Shougo) [inline src/nvim/edit.c]: Well, it seems different means.

**Comment by Shougo** [comment]
Fixed for the review and added the tests.

@justinmk Please check it.

**Comment by splinterofchaos** [comment]
LGTM :+1:

**Comment by Shougo** [comment]
Thanks.

**Comment by Shougo** [comment]
@justinmk I will change the title to `[RDY]` after your review.

**Comment by Shougo** [comment]
No comment is this patch?

**Comment by fmoralesc** [comment]
@Shougo I think you should mark it as RDY, it looks like those who have reviewed it have no further input to add. LGTM too.

**Comment by Shougo** [comment]
Thanks.  I will mark it as [RDY].

**Comment by Pyrohh** [comment]
Merged, thanks @Shougo and everyone who helped.

I only made this tiny change to align the new paragraphs up with the ones above:
```diff
diff --git a/runtime/doc/options.txt b/runtime/doc/options.txt
index 817cee0..08783eb 100644
--- a/runtime/doc/options.txt
+++ b/runtime/doc/options.txt
@@ -1609,11 +1609,11 @@ A jump table for the options with a short description can be found at |Q_op|.
 		    completion in the preview window.  Only works in
 		    combination with "menu" or "menuone".
 
-	   noinsert  Do not insert any text for a match until the user selects
+	  noinsert  Do not insert any text for a match until the user selects
 		    a match from the menu. Only works in combination with
 		    "menu" or "menuone". No effect if "longest" is present.
 
-	   noselect  Do not select a match in the menu, force the user to
+	  noselect  Do not select a match in the menu, force the user to
 		    select one from the menu. Only works in combination with
 		    "menu" or "menuone".
 
```

**Comment by Shougo** [comment]
Thank you!

**Pull request closed without merging**